File: Array--Medium/AllPermutations.cpp
#include<bits/stdc++.h>
using namespace std;

// Finding All the permutaitons using Recursion TC : O(N!*N), SC : O(1)
void Rec_allpermutations(int index, vector<int> & nums, vector<vector<int>>& res ) {
  //if index becomes equals to the size of nums then push the result and return.
  if(index==nums.size()) {
    res.push_back(nums);
    return ;
  }
  else {
    for(int i=index; i<nums.size(); i++) {
     // swaping index and i
      swap(nums[index], nums[i]);
      // call recursion for next
      Rec_allpermutations(index+1, nums, res);
      // again swap for next
      swap(nums[index], nums[i]);
    }
  }
}

File: Array--Medium/AllPermutations.js
// Time Complexity: O(N!*N), Space Complexity: O(1)
const function recAllPermutations(index, nums, res) {
  // If index becomes equal to the size of nums, push the result and return.
  if (index === nums.length) {
    res.push([...nums])
    return
  } else {
    for (let i = index; i < nums.length; i++) {
      ;[nums[index], nums[i]] = [nums[i], nums[index]]
      // Call recursion for next
      recAllPermutations(index + 1, nums, res)
      // Again swap for next
      ;[nums[index], nums[i]] = [nums[i], nums[index]]
    }
  }
}

const result = []
const nums = [1, 2, 3]

recAllPermutations(0, nums, result)
console.log(result)


